library/alloc/src/str.rs

```diff
@@ -615,50 +603,86 @@ pub unsafe fn from_boxed_utf8_unchecked(v: Box<[u8]>) -> Box<str> {
+///
+/// Returns a tuple of the converted prefix and the remainder starting from
+/// the first non-ascii character.
 #[inline]
 #[cfg(not(test))]
 #[cfg(not(no_global_oom_handling))]
-fn convert_while_ascii(b: &[u8], convert: fn(&u8) -> u8) -> Vec<u8> {
-    let mut out = Vec::with_capacity(b.len());
-
+fn convert_while_ascii(s: &str, convert: fn(&u8) -> u8) -> (String, &str) {
     const USIZE_SIZE: usize = mem::size_of::<usize>();
     const MAGIC_UNROLL: usize = 2;
     const N: usize = USIZE_SIZE * MAGIC_UNROLL;
-    const NONASCII_MASK: usize = usize::from_ne_bytes([0x80; USIZE_SIZE]);
 
-    let mut i = 0;
-    unsafe {
-        while i + N <= b.len() {
-            // Safety: we have checks the sizes `b` and `out` to know that our
-            let in_chunk = b.get_unchecked(i..i + N);
-            let out_chunk = out.spare_capacity_mut().get_unchecked_mut(i..i + N);
-
-            let mut bits = 0;
-            for j in 0..MAGIC_UNROLL {
-                // read the bytes 1 usize at a time (unaligned since we haven't checked the alignment)
-                // safety: in_chunk is valid bytes in the range
-                bits |= in_chunk.as_ptr().cast::<usize>().add(j).read_unaligned();
-            }
-            // if our chunks aren't ascii, then return only the prior bytes as init
-            if bits & NONASCII_MASK != 0 {
-                break;
-            }
+    let mut slice = s.as_bytes();
+    let mut out = Vec::with_capacity(slice.len());
+    let mut out_slice = out.spare_capacity_mut();
 
-            // perform the case conversions on N bytes (gets heavily autovec'd)
-            for j in 0..N {
-                // safety: in_chunk and out_chunk is valid bytes in the range
-                let out = out_chunk.get_unchecked_mut(j);
-                out.write(convert(in_chunk.get_unchecked(j)));
-            }
+    let mut i = 0_usize;
 
-            // mark these bytes as initialised
-            i += N;
+    // process the input in chunks to enable auto-vectorization
+    let mut is_ascii = [false; N];
+    while slice.len() >= N {
+        // Safety: out_slice was allocated with same lengths as input slice and gets updated with
+        // the same offsets
+        unsafe {
+            core::intrinsics::assume(slice.len() == out_slice.len());
         }
-        out.set_len(i);
+
+        let chunk = &slice[..N];
+
+        for j in 0..N {
+            is_ascii[j] = chunk[j] <= 127;
+        }
+
+        // auto-vectorization for this check is a bit fragile,
+        // sum and comparing against the chunk size gives the best result,
+        // specifically a pmovmsk instruction on x86.
+        if is_ascii.iter().map(|x| *x as u8).sum::<u8>() as usize != N {
+            break;
+        }
+
+        for j in 0..N {
+            out_slice[j] = MaybeUninit::new(convert(&chunk[j]));
+        }
+
+        i += N;
+        slice = &slice[N..];
+        out_slice = &mut out_slice[N..];
```

Comment:
Note that subslicing has to update two parts, the pointer and the length. I don't know if LLVM is clever enough to simplify it.

Maybe try using `slice::as_chunks` instead since that returns array-references which don't have to carry the length. Just using `i` and unchecked accesses would be another option, if necessary.

Reply:
LLVM seems to handle this quite fine. I tried a version with `array_chunks::<N>()` and `for (chunk, out_chunk) in chunks.zip(out_chunks)`, which looked nicer, but I could not get the remainder loop to work since that moves the iterators. There was also no measurable performance difference between the two versions without the scalar remainder loop.

Another version with slicing by `i..i+N` seems to break autovectorization again.

I find it difficult analysing this deeper, putting the `convert_while_ascii` into compiler explorer in isolation seems to optimize slightly better or differently than the full `to_lowercase` function. For example I see this sequence in the benchmark with `perf record`, which I think indicates llvm thinks something could alias the input ptr, but can't reproduce that in a smaller example:
```asm
mov          rsi,QWORD PTR [rbp-0x68]
vmovdqu      xmm3,XMMWORD PTR [rsi+r13*1]
```

I would prefer to stick with the slicing for now, but will update the PR with correctly slicing the spare capacity initially, and also replace the `assume` with `get_unchecked` since its shorter and should create less IR.